### PR TITLE
AssetBuilder - Fix `testInvalid()` failure. Switch to JWT.

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

This fixes a test-failure that recently appeared in `AssetBuilder`. Additionally, it switches the parameter-validation to JWT (for better security checks). For sites that lack the JWT pre-requisites (ie `CIVICRM_SIGN_KEYS`), it provides a transitional key.

Before+After
----------------------------------------

Based on the asset parameters, there is a digest-code. The usage of this digest-code evolves as follows:

| When | Description |
| -- | -- | 
| Before<br/> (Circa 5.56.0) | Digests are used for file-names and caching. Parameters (for the on-demand/debug variants of assets) are not validated. |
| Before<br/> (Circa 5.57.0) | Digests are used for file-names, caching, *and* parameter validation.
| After<br/> (This Patch) | Digests are used for file-names and caching. JWT does parameter validation. |

Technical Details: Background
-----------------------------

`AssetBuilder` was recently (5.57.0) updated to validate the digest-code in asset URLs (*for the on-demand/debug variants of assets*). But this led to new failures in `E2E\Core\AssetBuilderTest`.  Why?

The digest code relies on a semi-secret value, the runtime ID.   In this specific test (`E2E\Core\AssetBuilderTest::testInvalid()`), there are two processes. The PHPUnit process generates one digest-code (with one runtime ID) and sends it to the HTTP server -- but the HTTP server expects a different digest-code (with a different runtime ID).

Is it good or bad that phpunit+httpd have different runtime IDs? 🤷 Is it good or bad to use runtime ID here? I think runtime ID is fair as part of the cache identifier. But I don't think it's ideal for validating the origin of parameters. We have another tool (JWT) which is better designed for that.

Technical Details: This Patch
-------------------------------

This patch restores the use of `digest` code (and runtime ID) -- they are only used for purposes of file-names and caching. They no longer influence the validation-check. Instead, we use JWT.

There is one catch about JWT, -- as observed in the discussion around #25285, JWT requires a signing-key, and some upgraded sites may not have the signing-key configured. Given the centrality of asset-builder in the upgrade-workflow (e.g. for `civicrm/a/#/status`), we need a fallback so that site-builders can apply the upgrade and do the configuration. This defines a very limited fallback.

### CODE REVIEW

File: Civi/Crypto/CryptoRegistry.php [hunk 1]
```diff
@@ -84,6 +84,31 @@ public static function createDefaultRegistry(): CryptoRegistry {
...
+            // so we'll toss in more spaghetti. (At a minimum, this should mitigate bots/crawlers.)
+            \CRM_Utils_Constant::value('CIVICRM_DSN'),
+            \CRM_Utils_Constant::value('CIVICRM_UF_DSN'),
+            \CRM_Utils_Constant::value('CIVICRM_SITE_KEY') ?: $GLOBALS['civicrm_root'],
+            \CRM_Utils_Constant::value('CIVICRM_UF_BASEURL'),
+            \CRM_Utils_Constant::value('CIVICRM_DB_CACHE_PASSWORD'),
+            \CRM_Utils_System::getSiteID(),
+            \CRM_Utils_System::version(),
+            \CRM_Core_Config::singleton()->userSystem->getVersion(),
+            \CRM_Utils_Array::value('HTTP_HOST', $_SERVER, ''),
```
Comment (demeritcowboy) [inline Civi/Crypto/CryptoRegistry.php:107]: `\CRM_Utils_Array::value`

Ahhhhhhhhhhhhhhhhhhhhh.

Comment (totten) [reply Civi/Crypto/CryptoRegistry.php:107]: An oldie but goodie??

Comment (demeritcowboy) [reply Civi/Crypto/CryptoRegistry.php:107]: Oldie, yes. Can you use `$_SERVER['HTTP_HOST'] ?? ''`

Comment (totten) [reply Civi/Crypto/CryptoRegistry.php]: Sure, done.

**Comment by totten** [comment]
civibot, test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Pull request merged**